### PR TITLE
Add an example notebook for training an autoencoder using branches and Kino

### PR DESCRIPTION
**Pull request by hanrelan**
I added an example showing a workflow for how someone might train a model using Livebook and Axon. The two main contributions that I wasn't able to find in the other examples are using `Kino` to visualize the outputs and using Livebook branches so you can stop execution early without losing your model's state.

Would love to know if there's a better way to do anything here! Specifically - is using `Task.start` a good way to create the interactive Livebook cell? And is there a better workflow than using branches?

I was also unsure if this should go in notebooks/, examples/ or guides/

### CODE REVIEW

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+  {:nx, "~> 0.3.0"},
+  {:axon, "~> 0.2.0"},
+  {:req, "~> 0.3.1"},
+  {:kino, "~> 0.6.2"}
+])
+```
+
+## Intro
+
+The goal of this notebook is to build a Autoencoder from scratch using Livebook. This notebook is based on [Training an Autoencoder on Fashion MNIST](https://github.com/hanrelan/axon/blob/main/notebooks/generative/fashionmnist_autoencoder.livemd), but includes some tips on using Livebook to train the model and using [Kino](https://hexdocs.pm/kino/Kino.html) (Livebook's interactive widget library) to play with and visualize our results.
```
Comment (seanmor5) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:15]:
```suggestion
The goal of this notebook is to build an Autoencoder from scratch using Livebook. This notebook is based on [Training an Autoencoder on Fashion MNIST](https://github.com/hanrelan/axon/blob/main/notebooks/generative/fashionmnist_autoencoder.livemd), but includes some tips on using Livebook to train the model and using [Kino](https://hexdocs.pm/kino/Kino.html) (Livebook's interactive widget library) to play with and visualize our results.
```

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+## Data loading
+
+An autoencoder learns to recreate it's seen in its dataset. For this notebook, we're going to try something simple: generating images of digits using the MNIST digit recognition dataset.
```
Comment (seanmor5) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:19]:
```suggestion
An autoencoder learns to recreate data it's seen in the dataset. For this notebook, we're going to try something simple: generating images of digits using the MNIST digit recognition dataset.
```

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+```elixir
+mnist_base_url = "https://storage.googleapis.com/cvdf-datasets/mnist/"
+%{body: train_images} = Req.get!(mnist_base_url <> "train-images-idx3-ubyte.gz")
+# We're not going to use the labels so we won't grab them
+# %{body: train_labels} = Req.get!(base_url <> "train-labels-idx1-ubyte.gz")
+
+<<_::32, n_images::32, n_rows::32, n_cols::32, images::binary>> = train_images
+# Ignore the labels
+# <<_::32, n_labels::32, labels::binary>> = train_labels
+```
```
Comment (seanmor5) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:34]: Both this example and the one linked should probably use `SciData` since that has all of this logic built in :)

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+```elixir
+# Download and prepare the test set
+mnist_base_url = "https://storage.googleapis.com/cvdf-datasets/mnist/"
+%{body: test_images} = Req.get!(mnist_base_url <> "t10k-images-idx3-ubyte.gz")
+# I'm not going to use the labels so we won't grab them
+# %{body: train_labels} = Req.get!(base_url <> "train-labels-idx1-ubyte.gz")
+
+<<_::32, n_images::32, n_rows::32, n_cols::32, test_images::binary>> = test_images
```
Comment (seanmor5) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:69]: Same comment as above, I would look at using SciData for this task instead

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+  |> Nx.divide(255)
+
+test_images[[images: 0..2]] |> Nx.to_heatmap()
+```
+
+## Building the model
+
+An autoencoder is a a network that has the same sized input as output, with a "bottleneck" layer in the middle with far fewer parameters than the input. Its goal is to force the output to reconstruct the input. The bottleneck layer forces the network to learn a compressed representation of the space.
+
+The part of the autoencder that takes the input and compresses it into the bottleneck layer is called the *encoder* and the part that takes the compressed representation and reconstructs the input is called the *decoder*. Usually the decoder mirrors the encoder.
```
Comment (seanmor5) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:86]:
```suggestion
The part of the autoencoder that takes the input and compresses it into the bottleneck layer is called the *encoder* and the part that takes the compressed representation and reconstructs the input is called the *decoder*. Usually the decoder mirrors the encoder.
```

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+  )
+
+Kino.render(form)
+
+task =
+  Task.start(fn ->
+    for %{data: %{test_image_index: image_index}} <- Kino.Control.stream(form) do
+      test_image = test_images[[images: image_index]]
+      # [0] removes the batch dimension
+      reconstructed_image = Axon.predict(model, params, test_image)[0]
```
Comment (seanmor5) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:203]:
```suggestion
      reconstructed_image =
        model
        |> Axon.predict(params, test_image)
        |> Nx.squeeze()
```

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+It'd be nice to see how the model improves as it trains. In this section (also a branch since I plan to experiment and don't want to lose the execution state) we'll improve the training loop to use `Kino` to show us how it's doing.
+
+[Axon.Loop.handle](https://hexdocs.pm/axon/Axon.Loop.html#handle/4) gives us a hook into various points of the training loop. We'll can use it with the `:iteration_completed` event to get a copy of the state of the params after some number of completed iterations of the training loop. By using those params to render an image in the test set, we can get a live view of the autoencoder learning to reconstruct its inputs.
+
+```elixir
+# A helper function to display the input and output side by side
+combined_input_output = fn params, image_index ->
+  test_image = test_images[[images: image_index]]
+  reconstructed_image = Axon.predict(model, params, test_image)[0]
```
Comment (seanmor5) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:232]:
```suggestion
  reconstructed_image = Axon.predict(model, params, test_image) |> Nx.squeeze()
```

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+  Nx.concatenate([test_image, reconstructed_image], axis: :width)
+end
+
+Nx.to_heatmap(combined_input_output.(params, 0))
+```
+
+```elixir
+# Once again we'll use a Kino.Frame for a dynamically updating output
```
Comment (seanmor5) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:240]:
```suggestion

Once again we'll use `Kino.Frame` for dynamically updating output:

```elixir

```

**Review by seanmor5 — COMMENTED** [review]
Nice work and thank you @hanrelan ! I think using a task here is totally fine, @jonatanklosko and @josevalim might have different suggestions though as they're the more experienced ones with Kino here. 

As far as this example goes, do you think you can make the autoencoder a denoising autoencoder? The reason I ask is that will add another element/application to this which is explicitly different from the Fashion MNIST example. I think all you'd need to do is adjust your dataset to add some small gaussian noise to the inputs. There's a script example here: https://github.com/elixir-nx/axon/blob/main/examples/vision/cnn_image_denoising.exs

Otherwise I think this is great and a great demonstration of writing custom event handlers and using Kino/Livebook :)

**Comment by hanrelan** [comment]
@seanmor5 included your suggestions and updated it to be a denoising autoencoder.

One thing I noticed is that the `add_noise` function takes a really long time to execute on the train/test data. I think this is more an issue with `Nx` though, where `Nx.to_batched` needs to also be able to accept a `Stream` or something like that. I might take a look and see if I can contribute something there.

Next up I'm going to try putting together a VAE example which should demo custom layers and splitting up models :crossed_fingers:

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+      test_image_index: Kino.Input.number("Test Image Index", default: 0)
+    ],
+    submit: "Render"
+  )
+
+Kino.render(form)
+
+task =
+  Task.start(fn ->
+    for %{data: %{test_image_index: image_index}} <- Kino.Control.stream(form) do
```
Comment (jonatanklosko) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:240]: If you want a non-blocking version you can use `Kino.Control.stream/2` which starts a task under the hood :)

Note: the correct way to start a task in Livebook is `Kino.start_child({Task, fn -> ... end})`, this way it is cleaned up when the cell is reevaluated.

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+```elixir
+frame = Kino.Frame.new() |> Kino.render()
+
+form =
+  Kino.Control.form(
+    [
+      test_image_index: Kino.Input.number("Test Image Index", default: 0)
+    ],
+    submit: "Render"
+  )
```
Comment (jonatanklosko) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:234]: Idea: we could also just render a "Next" button and use `Kino.Control.stream/3` keeping track of the current index in state and rendering subsequent images on click. This would make the iteration quicker, but if we want to look at specific images, an input makes sense, so whichever you like more!

**Comment by seanmor5** [comment]
Nice! Do you want to merge this before the VAE? Or do you want to include the VAE here? I have no preference, although I think the VAE should be a separate notebook

**Comment by jonatanklosko** [comment]
@hanrelan cool stuff! Just a couple suggestions from me :)

File: notebooks/generative/mnist_autoencoder_using_kino.livemd (hunk 10 above)
Comment (hanrelan) [reply notebooks/generative/mnist_autoencoder_using_kino.livemd:234]: The input seems like a common use case so I'll leave it that way, but maybe I'll use the stateful version in the VAE notebook since it'll be useful to see the steps of the interpolation. Thanks for the pointer!

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+      test_image = test_images[[images: image_index]] |> add_noise.()
+
+      reconstructed_image =
+        model
+        |> Axon.predict(params, test_image)
+        # Get rid of the batch dimension
+        |> Nx.squeeze(axes: [0])
+
+      combined_image = Nx.concatenate([test_image, reconstructed_image], axis: :width)
+      Kino.Frame.render(frame, Nx.to_heatmap(combined_image))
```
[the remark this answers is not in the document]
Comment (hanrelan) [reply notebooks/generative/mnist_autoencoder_using_kino.livemd:250]: this is great! I was wondering how to do this

**Comment by hanrelan** [comment]
Thanks for the feedback @jonatanklosko @seanmor5 
I incorporated all of it and the image outputs look great now.

image

@seanmor5 I'll do the VAE in another notebook so it doesn't become too large. So I think this is ready to merge now

File: notebooks/generative/mnist_autoencoder_using_kino.livemd [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+  {:req, "~> 0.3.1"},
+  {:kino, "~> 0.6.2"},
+  {:scidata, "~> 0.1.9"},
+  {:stb_image, "~> 0.5.2"}
+])
+```
+
+## Introduction
+
+The goal of this notebook is to build a Denoising Autoencoder from scratch using Livebook. This notebook is based on [Training an Autoencoder on Fashion MNIST](https://github.com/hanrelan/axon/blob/main/notebooks/generative/fashionmnist_autoencoder.livemd), but includes some tips on using Livebook to train the model and using [Kino](https://hexdocs.pm/kino/Kino.html) (Livebook's interactive widget library) to play with and visualize our results.
```
Comment (jonatanklosko) [inline notebooks/generative/mnist_autoencoder_using_kino.livemd:17]:
```suggestion
The goal of this notebook is to build a Denoising Autoencoder from scratch using Livebook. This notebook is based on [Training an Autoencoder on Fashion MNIST](fashionmnist_autoencoder.livemd), but includes some tips on using Livebook to train the model and using [Kino](https://hexdocs.pm/kino/Kino.html) (Livebook's interactive widget library) to play with and visualize our results.
```

**Review by jonatanklosko — APPROVED** [review]
